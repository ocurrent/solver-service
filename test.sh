#!/bin/sh 

eval $(opam env)

#host=c2-4.equinix.ci.dev
host=arm64-jade-1.equinix.ci.dev

for JOBS in 1 $(seq 8 16 480) ; do
min=$(( JOBS < 160 ? JOBS : 160 ))
for INSTANCES in 1 $(seq 8 8 $min) ; do
for WORKERS in 1 $(seq 2 2 10) ; do

output="run-$INSTANCES-$WORKERS-$WORKERS-$JOBS"

echo Run $output

if ! [ -f $output ] ; then

rm *.service

for NAM in $(seq 1 $INSTANCES) ; do
sed "s/NAM/$NAM/g;s/CAP/$WORKERS/g;s/WOR/$WORKERS/g;" solver-service > solver-service-$NAM.service
done

ssh $host 'cd /etc/systemd/system ; for s in solver-service-* ; do systemctl stop $s ; done'
ssh $host rm /etc/systemd/system/solver-service-*
scp *.service $host:/etc/systemd/system
ssh $host systemctl daemon-reload
ssh $host 'cd /etc/systemd/system ; for s in solver-service-* ; do systemctl start $s ; done'

dune exec -- stress/stress.exe cluster ~/mtelvers.cap --count=$JOBS --solver-pool=solver-test > $output

fi

done
done
done

grep "Solved [0-9]* " run-* | awk ' { split($1, x, "[-:]") ; split($7, y, "(") ; print x[2] "\t" x[3] "\t" x[4] "\t" x[5] "\t" y[2] }' | sort -rg -k 5
