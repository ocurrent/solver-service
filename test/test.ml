open Eio.Std
open Utils

let () = Printexc.record_backtrace true

let debian_12_ocaml_5 =
  { Solver_service_api.Worker.Vars.
    arch = "x86_64";
    os = "linux";
    os_family = "debian";
    os_distribution = "debian";
    os_version = "12";
    ocaml_package = "ocaml-base-compiler";
    ocaml_version = "5.0";
    opam_version = "2.1.3";
    lower_bound = false;
  }

let platforms = [
  "debian-12-ocaml-5", debian_12_ocaml_5;
]

let test_simple t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", {| depends: [ "foo" ] |}] in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", {| depends: [ "ocaml-base-compiler" ] |};
  ]
  in
  solve t "Select foo.1.0" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages];
  solve t "Foo 1.1 now available" ~platforms ~root_pkgs ~commits:[
    opam_repo, (("foo.1.1", "") :: opam_packages)
  ];
  (* Doesn't do another git-fetch, and doesn't use latest commit. *)
  solve t "Retry with previous commit" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages];
  ()

let test_overlay t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let overlay_repo = Opam_repo.create "overlay.git" in
  let root_pkgs = ["app.dev", {| depends: [ "foo" ] |}] in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", {| depends: [ "ocaml-base-compiler" { >= "5.1" } ] |};
  ] in
  let overlay_packages = [
    "foo.1.0", {| depends: [ "ocaml-base-compiler" { = "5.0" } ] |};
  ] in
  solve t "Fails without overlay" ~platforms ~root_pkgs ~commits:[
    opam_repo, opam_packages;
  ];
  solve t "Selects 5.0 with overlay" ~platforms ~root_pkgs ~commits:[
    opam_repo, opam_packages;
    overlay_repo, overlay_packages;
  ];
  solve t "Overlay not needed" ~platforms ~root_pkgs:["app.dev", ""] ~commits:[
    opam_repo, opam_packages;
    overlay_repo, overlay_packages;
  ];
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", "";
  ] in
  let overlay_packages = [
    "foo.1.1", {| depends: [ "ocaml-base-compiler" { = "6.0" } ] |};
  ] in
  solve t "Non-overlaid versions still visible" ~platforms ~root_pkgs ~commits:[
    opam_repo, opam_packages;
    overlay_repo, overlay_packages;
  ]

let test_lower_bound t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", {| depends: [ "foo" ] |}] in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", "";
    "foo.1.1", "";
  ] in
  let platforms = platforms @ [
    "lower-bound", {debian_12_ocaml_5 with lower_bound = true}
  ] in
  solve t "Selects foo 1.0 for lower-bound" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages]

let test_no_solution t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", {| depends: [ "foo" ] |}] in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", {| depends: [ "ocaml-base-compiler" { = "6.0" } ] |};
  ] in
  solve t "No solution" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages]

let test_errors t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", {| syntax error |}] in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1.0", {| depends: [ "ocaml-base-compiler" ] |};
  ] in
  solve t "Invalid opam root pkg" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages];
  let root_pkgs = ["app.dev", ""] in
  let opam_packages_invalid = [
    "ocaml-base-compiler.5.0", "syntax error";
  ] in
  solve t "Invalid package in repo" ~platforms ~root_pkgs ~commits:[opam_repo, opam_packages_invalid];
  Fmt.pr "@.## Invalid commit@.@.";
  let req =
    { Solver_service_api.Worker.Solve_request.
      opam_repository_commits = [ "opam-repo.git", "17e66310c5c95560291a2d6499e7154d30f95f06"];
      root_pkgs = [];
      pinned_pkgs = [];
      platforms;
    }
  in
  let response = Solver_service.Solver.solve t ~log:stderr_log req in
  Fmt.pr "@[<v2>results:@,%a@]@." pp_response response;
  Fmt.pr "@.## Invalid repository@.@.";
  let req =
    { Solver_service_api.Worker.Solve_request.
      opam_repository_commits = [ "foo.git", "17e66310c5c95560291a2d6499e7154d30f95f06"];
      root_pkgs = [];
      pinned_pkgs = [];
      platforms;
    }
  in
  let response = Solver_service.Solver.solve t ~log:stderr_log req in
  Fmt.pr "@[<v2>results:@,%a@]@." pp_response response

let test_double_fetch t =
  Fmt.pr "@.";
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", Utils.add_opam_header ""] in
  let test req = Solver_service.Solver.solve t ~log:stderr_log req in
  let good_commit =
    Opam_repo.commit opam_repo [
      "ocaml-base-compiler.5.0", {|synopsis: "Force fetch"|};
    ]
  in
  let test_bad () =
    test {
      Solver_service_api.Worker.Solve_request.
      opam_repository_commits = ["opam-repo.git", "17e66310c5c95560291a2d6499e7154d30f9aaaa"];
      root_pkgs;
      pinned_pkgs = [];
      platforms;
    }
  and test_good () =
    test {
      Solver_service_api.Worker.Solve_request.
      opam_repository_commits = [good_commit];
      root_pkgs;
      pinned_pkgs = [];
      platforms;
    }
  in
  let both a b =
    let a, b = Fiber.pair a b in
    Fmt.pr "@[<v2>a:@,%a@]@." Utils.pp_response a;
    Fmt.pr "@[<v2>b:@,%a@]@." Utils.pp_response b
  in
  Fmt.pr "## Concurrent fetches of invalid commit@.@.";
  both test_bad test_bad;
  Fmt.pr "@.## One bad, one good@.@.";
  both test_bad test_good

let test_multiple_roots t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let opam_packages = [
    "ocaml-base-compiler.4.14", "";
    "ocaml-base-compiler.5.0", "";
    "ocaml.4.14", {| depends: [ "ocaml-base-compiler" {= "4.14"} ] |};
    "ocaml.5.0", {| depends: [ "ocaml-base-compiler" {= "5.0"} ] |};
  ] in
  let root_pkgs = [
    "api.dev", {| depends: [ "ocaml" ] |};
    "impl.dev", {| depends: [ "ocaml" {>= "5.0"} ] |};
  ] in
  solve t "Only compatible packages"
    ~root_pkgs
    ~commits:[opam_repo, opam_packages]
    ~platforms:[
      "ocaml-4.14", { debian_12_ocaml_5 with ocaml_version = "4.14" };
      "ocaml-5.0", debian_12_ocaml_5;
    ]

let test_pinned t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "foo.1", {| depends: [ "ocaml-base-compiler" {= "4.14"} ] |};
  ] in
  let root_pkgs = [ "root.dev", {| depends: [ "foo" ] |} ] in
  let pinned_pkgs = [
    "foo.2", {| depends: [ "ocaml-base-compiler" {= "5.0"} ] |};
  ] in
  solve t "Solution using pinned packages"
    ~root_pkgs
    ~pinned_pkgs
    ~commits:[opam_repo, opam_packages]
    ~platforms

let test_cancel t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
  ] in
  let root_pkgs = [ "foo.dev", "" ] in
  let cancelled = Promise.create_resolved () in
  solve t "Solution using pinned packages"
    ~cancelled
    ~root_pkgs
    ~commits:[opam_repo, opam_packages]
    ~platforms

let test_available t =
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
  ] in
  let root_pkgs = [
    "foo.dev", "";
    "foo-linux.dev", {| available: os = "linux" |};
  ] in
  solve t "foo-linux is only selected on Linux platform, but foo is selected for both"
    ~root_pkgs
    ~commits:[opam_repo, opam_packages]
    ~platforms:[
      "linux", debian_12_ocaml_5;
      "mac", { debian_12_ocaml_5 with os = "macos" };
    ]

let test_solve_cache t =
  let solve = solve_cache in
  let opam_repo = Opam_repo.create "opam-repo.git" in
  let root_pkgs = ["app.dev", {| depends: [ "foo" ] |}] in
  let depends = {| depends: [ "ocaml-base-compiler" "bar" ] |} in
  let opam_packages = [
    "ocaml-base-compiler.5.0", "";
    "bar.1.0", {| depends: [ "baz" ] |};
    "baz.1.0", "";
    "foo.1.0",depends;
    "foobar.0.1", "";
  ]
  in
  let first_opam_packages = opam_packages in
  let recent_commits =
    solve t "Select foo.1.0" ~platforms ~root_pkgs ~previous_commits:[opam_repo,[]]
      ~commits:[opam_repo, opam_packages]
  in
  let opam_packages = ("foo.1.1",depends) :: opam_packages in
  let recent_commits =
  solve t "Foo 1.1 now available (A direct dependency, the result will contain the new commit)" ~previous_commits:recent_commits ~platforms ~root_pkgs ~commits:[
    opam_repo, opam_packages
  ]
  in
  let opam_packages = ("foo.1.1",depends)::opam_packages in
  let recent_commits =
    solve t "Foo 1.1 again (hit the cache, the commit won't change)" ~previous_commits:recent_commits ~platforms ~root_pkgs ~commits:[
      opam_repo, opam_packages
    ]
  in
  let opam_packages =
   ("baz.1.0",{|depends: [ "foobar" ]|})
    ::(List.remove_assoc "baz.1.0" opam_packages)
  in
  let recent_commits =
    solve t "Baz 1.0 is a transitive dep of Foo, the cache will be invalidated(the result will contain the new commit)" ~previous_commits:recent_commits ~platforms ~root_pkgs ~commits:[
      opam_repo, opam_packages
    ]
  in

  let opam_packages = ("oof.1.0","") :: opam_packages in
  let recent_commits =
  solve t "Oof 1.0 now available (hit the cache, the commit won't change in the result)" ~previous_commits:recent_commits ~platforms ~root_pkgs ~commits:[
    opam_repo, opam_packages
  ]
  in
  solve t
    "Oof 1.1 now available (will invalidate the cache because foo 1.1 will be removed, the result will contain the new commit)"
    ~previous_commits:recent_commits ~platforms ~root_pkgs ~commits:[
    opam_repo, ("oof.1.1","") :: first_opam_packages
  ] |> ignore;
  ()

let () =
  Eio_main.run @@ fun env ->
  let domain_mgr = env#domain_mgr in
  let process_mgr = env#process_mgr in
  let cache_dir = "cache" in
  Lwt_eio.with_event_loop ~clock:env#clock @@ fun () ->
  Switch.run @@ fun sw ->
  let t = Solver_service.Solver.create ~sw ~domain_mgr ~process_mgr ~cache_dir ~n_workers:2 in
  [
    "Simple", test_simple;
    "Overlay", test_overlay;
    "Lower-bound", test_lower_bound;
    "No solution", test_no_solution;
    "Errors", test_errors;
    "Double fetch", test_double_fetch;
    "Multiple roots", test_multiple_roots;
    "Pinned", test_pinned;
    "Cancel", test_cancel;
    "Available", test_available;
    "Solve_cache", test_solve_cache;
  ]
  |> List.iter (fun (name, fn) ->
      Fmt.pr "@.# %s@." name;
      fn t
    );
  Fmt.epr "Tests successful!@."
