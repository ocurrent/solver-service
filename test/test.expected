
# Simple

## Select foo.1.0 ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, bf39e9df31e82a307e3daee9409f62d1a15acfe7)]
     lower_bound: false]

## Foo 1.1 now available ##

commits: [(opam-repo.git, [foo.1.1; ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit a400c503c1cd70e902c5387c3048b13df6345462
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.1; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, a400c503c1cd70e902c5387c3048b13df6345462)]
     lower_bound: false]

## Retry with previous commit ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, bf39e9df31e82a307e3daee9409f62d1a15acfe7)]
     lower_bound: false]

# Multiple OCaml versions

## Select foo.1.0 for 5.0 and 4.14.1 ##

commits: [(opam-repo.git,
           [ocaml-base-compiler.5.0; ocaml-base-compiler.4.14.1; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5; debian-12-ocaml-4]
Need to update opam-repo.git to get new commit a50e88bf106724a4758fdc7a9af58a2fd39d392f
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, a50e88bf106724a4758fdc7a9af58a2fd39d392f)]
     lower_bound: false;
   debian-12-ocaml-4:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.4.14.1]
     commits: [(opam-repo.git, a50e88bf106724a4758fdc7a9af58a2fd39d392f)]
     lower_bound: false]

# Overlay

## Fails without overlay ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit daf78f787d9147faf890684079654e3c74be5adf
results:
  []

## Selects 5.0 with overlay ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0]);
          (overlay.git, [foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, daf78f787d9147faf890684079654e3c74be5adf);
               (overlay.git, 98827377d66bfb42ca4d7d0bd17d3b59823cf3e3)]
     lower_bound: false]

## Overlay not needed ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0]);
          (overlay.git, [foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, daf78f787d9147faf890684079654e3c74be5adf)]
     lower_bound: false]

## Non-overlaid versions still visible ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0]);
          (overlay.git, [foo.1.1])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit d2fc42d092d44bea571d52cfcc4474428c9617f2
Need to update overlay.git to get new commit 4e63ed9c4f328b824c844e7e5eb56c564b970f69
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, d2fc42d092d44bea571d52cfcc4474428c9617f2)]
     lower_bound: false]

# Lower-bound

## Selects foo 1.0 for lower-bound ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0; foo.1.1])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5; lower-bound]
Need to update opam-repo.git to get new commit 54edcb0b76daa233fb597d4cc2cf9258a39ea49a
results:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.1; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, 54edcb0b76daa233fb597d4cc2cf9258a39ea49a)]
     lower_bound: false;
   lower-bound:
     compat_pkgs: [app.dev]
     packages: [app.dev; foo.1.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, 54edcb0b76daa233fb597d4cc2cf9258a39ea49a)]
     lower_bound: true]

# No solution

## No solution ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit 39b23821916e30bd19dbb6fb916c9248b41abd83
results:
  []

# Errors

## Invalid opam root pkg ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
results:
  Error: Error parsing app.dev: At ./<none>:4:10-4:15::
Parse error

## Invalid package in repo ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0])]
root_pkgs: [app.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit 17e66310c5c95560291a2d6499e7154d30f95f05
results:
  Error: Error parsing ocaml-base-compiler.5.0: At ./<none>:4:9-4:14::
Parse error

## Invalid commit

Need to update opam-repo.git to get new commit 17e66310c5c95560291a2d6499e7154d30f95f06
results:
  Error: Still missing commit after update!

## Invalid repository

results:
  Error: Eio.Io Process Child_error Exited (code 128),
  running command: git clone --bare foo.git cache/no_host/foo,
  cloning "foo.git"

# Double fetch

## Concurrent fetches of invalid commit

Need to update opam-repo.git to get new commit 17e66310c5c95560291a2d6499e7154d30f9aaaa
Need to update opam-repo.git to get new commit 17e66310c5c95560291a2d6499e7154d30f9aaaa
a:
  Error: Still missing commit after update!
b:
  Error: Still missing commit after update!

## One bad, one good

Need to update opam-repo.git to get new commit 17e66310c5c95560291a2d6499e7154d30f9aaaa
a:
  Error: Still missing commit after update!
b:
  [debian-12-ocaml-5:
     compat_pkgs: [app.dev]
     packages: [app.dev; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, b4f969c46abb646b30bbf2cf35f9175daeb21741)]
     lower_bound: false]

# Multiple roots

## Only compatible packages ##

commits: [(opam-repo.git,
           [ocaml-base-compiler.4.14; ocaml-base-compiler.5.0; ocaml.4.14;
            ocaml.5.0])]
root_pkgs: [api.dev; impl.dev]
platforms: [ocaml-4.14; ocaml-5.0]
Need to update opam-repo.git to get new commit 27e6f844ca9ba0a3cef689b91b240f0e18fc6315
results:
  [ocaml-4.14:
     compat_pkgs: [api.dev]
     packages: [api.dev; ocaml.4.14; ocaml-base-compiler.4.14]
     commits: [(opam-repo.git, 27e6f844ca9ba0a3cef689b91b240f0e18fc6315)]
     lower_bound: false;
   ocaml-5.0:
     compat_pkgs: [api.dev; impl.dev]
     packages: [api.dev; impl.dev; ocaml.5.0; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, 27e6f844ca9ba0a3cef689b91b240f0e18fc6315)]
     lower_bound: false]

# Pinned

## Solution using pinned packages ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0; foo.1])]
root_pkgs: [root.dev]
platforms: [debian-12-ocaml-5]
pinned: [foo.2]
Need to update opam-repo.git to get new commit 03bb2831c6015ac386adf820cc501c74d033320e
results:
  [debian-12-ocaml-5:
     compat_pkgs: [root.dev]
     packages: [foo.2; ocaml-base-compiler.5.0; root.dev]
     commits: [(opam-repo.git, 03bb2831c6015ac386adf820cc501c74d033320e)]
     lower_bound: false]

# Cancel

## Solution using pinned packages ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0])]
root_pkgs: [foo.dev]
platforms: [debian-12-ocaml-5]
Need to update opam-repo.git to get new commit 1bc28b8e8d98db6e524822c6f28bddebbc3504a3
results:
  Cancelled

# Available

## foo-linux is only selected on Linux platform, but foo is selected for both ##

commits: [(opam-repo.git, [ocaml-base-compiler.5.0])]
root_pkgs: [foo.dev; foo-linux.dev]
platforms: [linux; mac]
results:
  [linux:
     compat_pkgs: [foo.dev; foo-linux.dev]
     packages: [foo.dev; foo-linux.dev; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, 1bc28b8e8d98db6e524822c6f28bddebbc3504a3)]
     lower_bound: false;
   mac:
     compat_pkgs: [foo.dev]
     packages: [foo.dev; ocaml-base-compiler.5.0]
     commits: [(opam-repo.git, 1bc28b8e8d98db6e524822c6f28bddebbc3504a3)]
     lower_bound: false]
